### PR TITLE
[CLOUD-3289] adding optional module for transaction recovery jdbc driver

### PR DESCRIPTION
**Pull request by ochaloup**
https://issues.redhat.com/browse/CLOUD-3289

Adding a way to use non-postgresql and non-mysql JDBC drivers and configure the connection for the transaction recovery scaledown processing when JDBC object store is used.

This change, fixes and code additions and released jbosstm tools `.jars`, expects for user defining manually a module named as `io.narayana.openshift-recovery.jdbc` which is the place where JDBC driver jar files are stored and used by the recovery module `io.narayana.openshift-recovery` from.

The module may be created with galleon feature pack - e.g. I did it for oracle like this: https://github.com/ochaloup/wildfly-datasources-galleon-pack/commit/3d475603638016b6a21843f1832e148136ef9c71.

### CODE REVIEW

**Comment by ochaloup** [comment]
@jfdenise @kabir can you take a look, please?

It would be good to add a test for this too (the same as for https://github.com/wildfly/wildfly-cekit-modules/pull/142). I was looking for some already existing but I haven't found. What could be a good place to put them to? Thanks!

**Comment by jfdenise** [comment]
@ochaloup , for tests, you could define a new tests/features/tx-recovery.feature file. If you need to define a new test application (for s2i), then you can add it to tests/examples

**Comment by jfdenise** [comment]
@ochaloup , seems that your oracle driver could be a contribution to the project? What do you think?
I am possibly thinking at an evolution that would allow user to directly use their custom module. What do you think of treating the io.narayana.openshift-recovery.jdbc module as an alias?
1) new env var: CUSTOM_JDBC_MODULE=org.foo.bar
2) If CUSTOM_JDBC_MODULE is set, you create the io.narayana.openshift-recovery.jdbc module.xml and install it in JBOSS_HOME from your partitionPV.sh script.
Its content would be ultra simple:
`<module-alias` xmlns="urn:jboss:module:1.9" name="io.narayana.openshift-recovery.jdbc" target-name="<value of CUSTOM_JDBC_MODULE>"/>`

So any module installed with galleon, would be then directly usable.

**Comment by ochaloup** [comment]
@jfdenise ok, that's nice. I was not aware of the `module-alias`. That sounds much better than force the user to pack the JDBC driver to the module of the particular name. I will try to do so.

The PR for the galleon pack and oracle driver is here: https://github.com/wildfly-extras/wildfly-datasources-galleon-pack/pull/11
I'm not sure if the datasource part should be added or the driver is fine. With the driver I have few doubts.

**Comment by jfdenise** [comment]
@ochaloup , nice. Seems that, when we have oracle driver in wildfly-extras, we can add a tx-recovery test for it.
You can add a new test-app such as: https://github.com/jfdenise/jboss-eap-modules/tree/master/tests/examples/test-app-prov-mysql
Then add an s2i test similar to: https://github.com/jfdenise/jboss-eap-modules/blob/master/tests/features/driver.feature#L11

**Comment by ochaloup** [comment]
@jfdenise I updated the PR and added basic tests. May you review it, please?
Do you think this PR could be merged as it is?

I'm happy to add a test for oracle datasource as you mentioned. But I would add it as a separate task if you don't mind. For this could be done a new release of `wildfly-datasources-galleon-pack` with oracle ds available (https://github.com/wildfly-extras/wildfly-datasources-galleon-pack/pull/11).

**Comment by jfdenise** [comment]
@ochaloup , actually you could already add a test for it by adding a test-app similar to: https://github.com/jboss-container-images/jboss-eap-modules/tree/master/tests/examples/test-app-drivers (NB, you can't use this test-app as is, it creates the module org.postgresql that your narayana module is also optionally loading). But by simply doing some module renaming it should work.
Seems that it would be safer to have a test for this feature, specially if this goes into next release.

**Comment by ochaloup** [comment]
@jfdenise ok, I'm sorry I didn't get it. I thought you talk about general tests for oracle driver.
In scope of this PR I already added the tests which checks the changed behaviour.
Btw. this PR is about improving already existing functionality.

I added a check for the alias module creation as it's available for the behave python tests. And I think it addresses the pointig of using mock test-app with renamed module name. That's good point.
But still I don't know if there is something more that could be tested.
Especially I can't see a functionality that could be  tested with the mock test-app with renamed driver.

To be more specific here. The `partitionPV.sh`, when configured with env variable to run with JDBC connection, arranges the data to connect to database. It needs information about host, port and path to JDBC driver. Then there is used a java jar to immediately calls to database. It happens before the EAP is started.
The behave tests may verify that the all necessary information were provided. But there is no database running in the background. It's not possible to check if the driver class is loaded correctly - if there is a module providing the driver there is still not possible to check that connection of java jar is running right. We can't check if connection to Oracle DB succeeds either with fake renamed driver or even with real one. These tasks could be done only in the integration tests.

In summary the PR contains tests which verifies on env variables to be setup. What should be other scenarios to be added to this PR?

**Comment by jfdenise** [comment]
I think it is all set. I am approving it. I think @luck3y should also approve it. Thank-you.
